### PR TITLE
DOC: NDFrame fillna method add use case

### LINKED ISSUE #38917: Passing DataFrame of values to fillna()
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

I'm not sure if this is a bug, a documentation issue, or just user error.
I found related issues #9750 #29048 that help me get the desired output but still don't explain why behavior doesn't match expectation based on my reading of the docs.

#### Code Sample

Following #9750 here's some code to reproduce the issue. 

```python
df = pd.DataFrame([[1, 1, 1],[2, 1, 1],[2, 1, 1],
                   [np.nan, np.nan, np.nan]], columns=["a","b","c"])
df.fillna(df.mode())
```
The resulting DataFrame is just `df` itself, i.e.  without any null values filled. 

#### Problem description

I understand from the issues I mentioned (and numerous StackOverflow answers) that this is due to `df.mode()` returning a `DataFrame`, as opposed to a `Series`, (which is why e.g. `df.mean()` gives the expected results). To get the desired output `df.fillna(df.mode().iloc[0])` works in this case. 

But I'm confused by [the docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html) which describe the value parameter:

> value: scalar, dict, Series, or DataFrame

    Value to use to fill holes (e.g. 0), alternately a dict/Series/DataFrame of values 
    specifying which value to use for each index (for a Series) or column (for a 
    DataFrame). Values not in the dict/Series/DataFrame will not be filled. This value 
    cannot be a list.

From how I read this, you should be able to pass a DataFrame with the same column names and have values filled for those columns. 

Just in case the passed in `DataFrame` needed to be the same shape, I tried

```python
df2 = pd.DataFrame(np.zeros((4, 3)))
df.fillna(df2)
```
 But still got `df` back.

I read the source but the answer still eludes me.

#### Expected Output

The expected behavior is that the missing values will be filled with mode of each column.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : b5958ee1999e9aead1938c0bba2b674378807b3d
python           : 3.8.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.5
numpy            : 1.19.2
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 51.0.0.post20201207
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by aniaan**
- [x] closes #38917 
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

cc @MarcoGorelli

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -6362,6 +6360,16 @@ def fillna(
         1   3.0 4.0 NaN 1
         2   NaN 1.0 NaN 5
         3   NaN 3.0 NaN 4
+
+        Replace with a DataFrame with the same column name
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:6364]: I think this could be clarified - replacement happens along same column names and same indices. Maybe the example `df2` could also have a column with a different name, such as `columns=list("ABCE")` so it's clear that replacement only happens for the columns they have in common

File: pandas/core/generic.py [hunk 2]
```diff
@@ -148,7 +148,6 @@
     you to specify a location to update with some value.""",
 }
 
-
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:151]: why change this? could unrelated changes be reverted please?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (BeanNan) [reply pandas/core/generic.py]: You are correct, now corrected

File: pandas/core/generic.py [hunk 3]
```diff
@@ -6362,6 +6362,16 @@ def fillna(
         1   3.0 4.0 NaN 1
         2   NaN 1.0 NaN 5
         3   NaN 3.0 NaN 4
+
+        Replacement happens along same column names and same indices
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:6366]: Can you write this as "When replacing with a DataFrame, replacement happens along the same column names and same indices:" (may need to split it across two lines). Other than that, looks good, thanks!

Comment (BeanNan) [reply pandas/core/generic.py]: It has been updated, please take a look

**Review by jreback — APPROVED** [review]
lgtm. @MarcoGorelli merge when satisfied

File: pandas/core/generic.py [hunk 4]
```diff
@@ -6362,6 +6362,17 @@ def fillna(
         1   3.0 4.0 NaN 1
         2   NaN 1.0 NaN 5
         3   NaN 3.0 NaN 4
+
+        When filling using a DataFrame, replacement happens along
+        the same column names and same indices
+
+        >>> df2 = pd.DataFrame(np.zeros((5, 5)), columns=list("ABCED"))
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:6369]: could `df2` have columns `"ABCE"`, so it's clear that there will be no replacement in column D (which is absent from `df2`)?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 4 above)
Comment (BeanNan) [reply pandas/core/generic.py]: Changed

**Comment by jreback** [comment]
thanks @BeanNan
